### PR TITLE
Add support for Snowflake Language

### PR DESCRIPTION
**Pull request by primeapple**
This is first draft to solve #366 . It does not work yet, mainly because I had some troubles with the `QuoteTypes`.

## Review
Please review, have a look at the following things:

1. `TODO` comments show lines where I'm totally lost
2. I'm very unsure about the `reservedComments`, the `functions` and the `keywords`.
3.  Did I put the correct things in there, or do they belong somewhere else?
4. Please check the places where I used the `[thing] {a|b|c} ...` syntax, if I used them correctly.
5. Did I miss any tests?
6. Any other places that have to be changed?

## Questions
I want to support accessing elements in unstructured data.
This can be done via square brackets `[,]` or via colons `:` , like [here](https://docs.snowflake.com/en/user-guide/querying-semistructured.html#bracket-notation):
```sql
select src:salesperson.name from car_sales

select src['salesperson']['name'] from car_sales
```
How should I do this? I added `:` as an operator, is this right? How about the brackets?

Thanks for this Repo!

### CODE REVIEW

**Comment by nene** [comment]
Thanks for the effort.

#### Regarding your main question about colon and square-bracket syntax.

- For brackets, adding `extraParens: ['[]']` should do the trick. See for example in `n1ql.formatter.ts`
- With colon, things are a bit trickier. You should indeed add it as an operator. However, currently the default is to format colon with a space after (as in `name: value`). To fix this, a language-specific condition should be added to `ExpressionFormatter.formatOperator()` - there already is a similar thing for `@` operator in PL/SQL.

#### In general

It would be really helpful if you could also update our [wiki](https://github.com/sql-formatter-org/sql-formatter/wiki) with information about Snowflake SQL syntax. Otherwise I'll have to do all this research on my own before I can merge this PR.

**Comment by primeapple** [comment]
I will do all the stuff you mentioned.
Please go other the `TODO` comments in the changes. I can't run tests because I don't understand the Quote types yet...

File: src/languages/snowflake/snowflake.formatter.ts
Comment (nene) [inline src/languages/snowflake/snowflake.formatter.ts:19]: This generated list is not really particularly useful. For example `INSERT` shouldn't be listed as a separate command, instead there should be `INSERT [OVERWRITE] INTO`.

I suggest you instead look at the `reservedCommands` in `sql.formatter.ts` and use it as a base and modify it to match Snowflake syntax. Also, it would be nice to keep things in the same order (e.g. starting with `WITH`, `FROM`, `WHERE`), so it's easier to compare with other other dialect definitions in *.formatter.ts files.

File: src/languages/snowflake/snowflake.formatter.ts
Comment (nene) [inline src/languages/snowflake/snowflake.formatter.ts:256]: As I understand, backticks (`` ` ``) aren't actually part of Snowflake SQL syntax (the docs suggest their usage inside JavaScript when dealing with SQL strings).

**Comment by nene** [comment]
Regarding quote types, see: https://github.com/sql-formatter-org/sql-formatter/blob/master/src/lexer/regexFactory.ts#L99

The `-qq` and `-bs` suffixes at the end of quote types define what sort of escaping (if any) is used inside the quotes.

**Comment by nene** [comment]
Some comments about wiki updates:

- [Parameters page](https://github.com/sql-formatter-org/sql-formatter/wiki/parameters) says that Snowflake supports `$1` parameters and links to [select docs](https://docs.snowflake.com/en/sql-reference/sql/select.html#parameters). I'm pretty sure these are something else than parameters one would use in prepared statements. I think these things are more similar to variables - they also share a very similar syntax. It's probably better to reference this syntax from [Variables page](https://github.com/sql-formatter-org/sql-formatter/wiki/variables). Probably the simplest path for supporting this syntax is also to treat them the same as variables - at lest that would be my first gut-instinct-guess.

- [Comments page](https://github.com/sql-formatter-org/sql-formatter/wiki/comments) mentions support for `# comments` but says these aren't actually supported. Syntax highlighting really shouldn't be used as a guide for what is and isn't supported. So probably best to be removed. Too bad you haven't found any official documentation for the comment syntax. Found one [Knowledge Base page](https://community.snowflake.com/s/article/how-to-comment-or-uncomment-multiple-lines-in-the-sql-worksheet) that at least mentions the `//` syntax - we could link to that.

**Comment by primeapple** [comment]
I updated all the wiki pages for snowflake (also updated your suggestions).
One problem is the [Window-Clause](https://github.com/sql-formatter-org/sql-formatter/wiki/WINDOW-clause). Snowflake seems to not have support for a `WINDOW` clause in it's [Query Syntax)(https://docs.snowflake.com/en/sql-reference/constructs.html). There is [QUALIFY](https://docs.snowflake.com/en/sql-reference/constructs/qualify.html) but I'm unsure if this is the same?  
However, there is support for [Window Functions](https://docs.snowflake.com/en/sql-reference/functions-analytic.html) but also here I'm unsure if it is the same as in the other languages.  
Could you have a look into this?

I will rewriting the code now.

**Comment by nene** [comment]
Thanks. Updated the docs for Snowflake window functions syntax.

**Comment by nene** [comment]
BTW. I would appreciate if you didn't do further force-pushes to this branch. They make it hard to tell what has changed.

File: src/languages/snowflake/snowflake.formatter.ts
Comment (primeapple) [reply src/languages/snowflake/snowflake.formatter.ts:19]: Ok, I did this like in the `postgresql.formatter.ts`. I added the basic commands with the `{..} [...]` syntax, as well as the generated List because it contains many many useful phrases.

**Comment by primeapple** [comment]
First refactor done.
Please the `snowflake.test` shows an infinity loop. Could you have a look into it? It happens even with a single test case.

File: src/languages/snowflake/snowflake.functions.ts
Comment (nene) [inline src/languages/snowflake/snowflake.functions.ts:108]: operators shouldn't be in functions list

File: src/languages/snowflake/snowflake.functions.ts
Comment (nene) [inline src/languages/snowflake/snowflake.functions.ts:122]: operators shouldn't be in functions list

File: src/languages/snowflake/snowflake.functions.ts
Comment (nene) [inline src/languages/snowflake/snowflake.functions.ts:201]: operators shouldn't be in functions list

File: src/languages/snowflake/snowflake.functions.ts
Comment (nene) [inline src/languages/snowflake/snowflake.functions.ts:235]: This `[ ... ]` syntax won't work in here because this list of strings is not wrapped inside `expandPhrases()` call.

Besides, these are again operators, not function names. That it, you won't be calling them as `IS NOT NULL(value)`

File: src/languages/snowflake/snowflake.functions.ts
Comment (nene) [inline src/languages/snowflake/snowflake.functions.ts:518]: Currently these won't work because the functions list gets concatenated to a regex and `$` will attempt to match end-of-line.

I should fix this by adding escaping of function names before constructing a regex.

**Review by nene — COMMENTED** [review]
The reason for the infinite loop is `||` operator in functions list, which gets concatenated into a long regex containing all function names and ends up matching empty string.

**Review by nene — COMMENTED** [review]
Some additional problems that popped up to me

**Comment by nene** [comment]
Regarding the formatting of `:` operator. I made a change to make the customization of formatting simpler: #466

One should now be able to simply add this to SnowflakeFormatter class:

```ts
formatOptions() {
  return { alwaysDenseOperators: [':'] };
}
```

**Comment by nene** [comment]
Also added escaping of keywords, so the `$` character in function names should no more pose a problem: #473

File: src/languages/snowflake/snowflake.formatter.ts
Comment (primeapple) [reply src/languages/snowflake/snowflake.formatter.ts:330]: Changed `d*` to `\\d*` to support any number sequence

**Comment by primeapple** [comment]
I'm sorry, I screwed up the commits by including your commits into this PR. This is why I usually force push, it allows me to cleanly integrate commits to master with `rebase` to the feature branch and then force push the feature branch.

**Comment by primeapple** [comment]
Ok, I reverted your commits in this PR and instead merged master into this branch. Everything fine now. Puhhh...

The only new important commit for you is [8dc6f51](https://github.com/sql-formatter-org/sql-formatter/pull/454/commits/8dc6f518a2e0bf9d76458c6a24274795b14f5a3b)

File: test/snowflake.test.ts
Comment (primeapple) [inline test/snowflake.test.ts:73]: These tests fail, unsure why.

File: src/languages/snowflake/snowflake.formatter.ts
Comment (primeapple) [inline src/languages/snowflake/snowflake.formatter.ts:35]: Are multiple following `[...]` blocks supported? I'm unsure because of the spaces in between, I dont want it to become something like `INSERT  INTO` (2 spaces in between).

**Comment by primeapple** [comment]
1. Is there an option to define how identifiers are allowed to look like?
In Snowflake they can include any symbol when inside `""`, but without quoting they have a certain rule. There is the `identChars` Config option, but I'm unsure if this refers to quoted or unquoted identifiers.

2. Is there an option to test variables?

3. Did I miss any obvious `support...` tests?

File: src/languages/snowflake/snowflake.formatter.ts
Comment (primeapple) [inline src/languages/snowflake/snowflake.formatter.ts:330]: Are these correctly written?

1. Should be anything like `$number`, where `number > 0`
2. Shoud be anything like `$identifier`, where identifier starts with char `a-Z`, can contain any combination of multiple `$`, `a-Z` and `0-9`

File: test/snowflake.test.ts
Comment (nene) [reply test/snowflake.test.ts:73]: Ah yeah, I forgot that there's a special formatting rule about the `:` operator and so the `alwaysDenseOperators` won't work with `:`. I'll have to fix that.

However, the second test case surprising for me. Is it really the case that Snowflake does not allow spaces around `:` operator? In that case we'll have to treat it as part of identifier instead, otherwise extra whitespace always gets allowed.

File: src/languages/snowflake/snowflake.formatter.ts
Comment (nene) [reply src/languages/snowflake/snowflake.formatter.ts:35]: Yep that's fine. The spaces will be taken care of by `expandPhrases()`. In the end though any number of whitespace characters between these words will be allowed by the parser.

File: src/languages/snowflake/snowflake.formatter.ts
Comment (nene) [reply src/languages/snowflake/snowflake.formatter.ts:330]: Looks fine to me. I'd suggest simplifying the first one to just:

```
[$][0-9]+
```

It's OK if we accept SQL that's not fully correct - the goal is formatting, not validation. It would only pose a problem with `$0` would be some sort of special syntax that we'd like to handle elsewhere.

**Comment by nene** [comment]
> Is there an option to define how identifiers are allowed to look like?

Yep, there are two options for that:

- `identTypes` describes quoted identifiers (with all characters allowed inside quotes).
- `identChars` describes unquoted identifiers.

Both of these seem to be configured correctly.

> Is there an option to test variables?

No, the variable syntax differs quite a bit between SQL dialects. Have a look at the variable tests in `behavesLikeMariaDbFormatter.ts` and create some custom tests for Snowflake.

> Did I miss any obvious support... tests?

Doesn't seem so at first glance.

File: test/snowflake.test.ts
Comment (primeapple) [reply test/snowflake.test.ts:73]: I'll investigate this.

File: test/features/truncateTable.ts
Comment (nene) [reply test/features/truncateTable.ts:28]: Though this test too is not that useful. This syntax is only supported by Snowflake, and like elsewhere, testing all these syntax variations doesn't really add much value.

File: test/snowflake.test.ts
Comment (nene) [inline test/snowflake.test.ts:60]: We should instead test that $ character can be used inside unquoted identifiers. The quoted ones are already covered by `supportsIdentifiers()` test.

File: src/languages/snowflake/snowflake.formatter.ts
Comment (nene) [inline src/languages/snowflake/snowflake.formatter.ts:36]: This is not correct. As I understand, the syntax is:

```
WHEN condition THEN INTO tablename
ELSE INTO tablename
```

We could break this up into the following:

```
'WHEN',
'THEN INTO',
'ELSE',
```

File: test/snowflake.test.ts
Comment (primeapple) [reply test/snowflake.test.ts:73]: Wow, I did not know this. It does accept `:` with spaces around!

Comment (primeapple) [reply test/snowflake.test.ts:73]: I fixed these tests and added some for the bracket and dot notation as well. Now the only thing failing is the `:` operator since it gets succeeding whitespace.

File: src/languages/snowflake/snowflake.formatter.ts
Comment (primeapple) [reply src/languages/snowflake/snowflake.formatter.ts:311]: Snowflake does not support these.

File: test/features/schema.ts
Comment (primeapple) [reply test/features/schema.ts:5]: You're right, restored this.

File: test/features/createView.ts
Comment (primeapple) [reply test/features/createView.ts:62]: Do you want me to open an issue for this?

File: src/languages/snowflake/snowflake.formatter.ts
Comment (primeapple) [reply src/languages/snowflake/snowflake.formatter.ts:36]: Ahh, you are right. Good find.
We can just remove `WHEN` and write `[THEN | ELSE] INTO`

File: src/languages/snowflake/snowflake.formatter.ts
Comment (primeapple) [reply src/languages/snowflake/snowflake.formatter.ts:330]: I see your point, but why not just leave it? It's not that complicated, I can rewrite it to 
`[$][1-9][0-9]*`
if that's better readable?

**Comment by primeapple** [comment]
I understand your concerns regarding the tests.
However in my opinion they have the purpose of explaining the accepted syntax without having to look into the actual source files.

For example with `supportsTruncateTable(format, { withoutTable: true, ifExists: true })` we can very easily see all the dialects that support `TRUNCATE IF EXISTS`, without having to manually look into the source files?

We've 3 options:
1. Leave the tests in this PR, add the other languages that support them later (I can do this if you want?)
2. Add the tests completly in a later PR
3. Remove the tests

Option 3. would decrease test coverage and verbosity for the sake of a few less lines and some milisecons of test time.

You are the maintainer, it's up to you.

File: test/features/createView.ts
Comment (nene) [reply test/features/createView.ts:62]: No worries, I mainly just wrote this comment as a not for myself.

File: src/languages/snowflake/snowflake.formatter.ts
Comment (nene) [reply src/languages/snowflake/snowflake.formatter.ts:330]: Might as well just leave it as is.

**Comment by nene** [comment]
OK, let the tests be for now as they are. My main concern was with the `SET SCHEMA` test. Others are not that big of a deal.

**Comment by nene** [comment]
OK. I fixed the `:` handling. If you now merge in the `master`, it should work.

Though now you'll also need to add the `::` operator to the `alwaysDenseOperators` config.

**Comment by primeapple** [comment]
Everything done.
Is this mergable now? :smiley:

File: src/languages/snowflake/snowflake.formatter.ts
Comment (nene) [reply src/languages/snowflake/snowflake.formatter.ts:311]: Does it really not? I see them in [snowflake documentation.](https://docs.snowflake.com/en/sql-reference/sql/create-table-constraint.html)

Plus, to test these, we should include `supportsConstraints()` helper.

File: src/languages/snowflake/snowflake.formatter.ts
Comment (nene) [inline src/languages/snowflake/snowflake.formatter.ts:36]: Should instead use:
```suggestion
  '{THEN | ELSE} INTO',
```

File: test/snowflake.test.ts
Comment (nene) [inline test/snowflake.test.ts:78]: No need for this test, it's already covered by `supportsArrayAndMapAccessors()`

File: test/features/truncateTable.ts
Comment (nene) [reply test/features/truncateTable.ts:28]: Let's also fix this inconsistent indentation in here.

**Review by nene — COMMENTED** [review]
Found a few more issues.

**Comment by primeapple** [comment]
I somehow can't answer your comments, so I'll do it here:

I added to reservedPhrases:
```js
  'MATCH {FULL | SIMPLE | PARTIAL}',
  'ON {UPDATE | DELETE} {CASCADE | SET NULL | SET DEFAULT | RESTRICT | NO ACTION}',
  'INITIALLY {DEFERRED | IMMEDIATE}',
```
This breaks the `supportConstraints` tests, because we are doing `ON UPDATE CURRENT_TIMESTAMP` there.
This test is problematic anyway because we are using it for many languages that don`t even support this line (Postgresql for example).

I fixed the test, we can add the correct specifications for other languages later.

**Review by nene — APPROVED** [review]

**Comment by nene** [comment]
Thanks for putting up with all my complaints :)

I'll merge this in and create a beta release out of it.

**Comment by primeapple** [comment]
Thank you for all the fast feedback. It was my first open source contribution and you made the process very easy.

**Comment by nene** [comment]
It's now released as 11.0.0-beta.1

**Pull request merged**
